build.wake

```diff
@@ -0,0 +1,416 @@
+  def buildEnv = sdk.getFreedomESDKBuildEnv
+  def bspDir = "{sdkDir}/bsp/{targetName}"
+  def runDir = "{sdkDir}/bsp"
+  def updateTargets = "{runDir}/update-targets.sh"
+
+  def cmdline =
+    def venvPath = relative runDir (freedomESDKVenvDir buildEnv)
+    def updateTargetsScript = relative runDir updateTargets
+    def targetDTS = relative runDir coreDTS.getPathName
+    def sdkPath = relative runDir sdkDir
+    "bash", "-c", "%
+      export FREEDOM_E_SDK_VENV_PATH=$(pwd)/%{venvPath}
+      touch ${FREEDOM_E_SDK_VENV_PATH}/.stamp
+      $(pwd)/%{updateTargetsScript} \
+        --target-name %{targetName} \
+        --target-dts=%{targetDTS} \
+        --sdk-path=%{sdkPath} \
+        --target-type %{targetType.targetTypeToString}
+      %", Nil
+
+  def inputs =
+    mkdir bspDir,
+    coreDTS,
+    sdk.getFreedomESDKInstalledFiles
+    ++ (installFreedomESDKVirtualenv buildEnv).getFreedomESDKVirtualenvInstalledFiles
+
+  def fnOutputs _ =
+    def alwaysOutputs =
+      "{bspDir}/core.dts",
+      "{bspDir}/design.dts",
+      "{bspDir}/design.svd",
+      "{bspDir}/metal.default.lds",
+      "{bspDir}/metal.ramrodata.lds",
+      "{bspDir}/metal.scratchpad.lds",
+      "{bspDir}/metal.freertos.lds",
+      "{bspDir}/settings.mk",
+      Nil
+    match targetType
+      ArtyTargetType = "{bspDir}/openocd.cfg", alwaysOutputs
+      VC707TargetType = "{bspDir}/openocd.cfg", alwaysOutputs
+      VCU118TargetType = "{bspDir}/openocd.cfg", alwaysOutputs
+      HiFiveTargetType = "{bspDir}/openocd.cfg", alwaysOutputs
+      _ = alwaysOutputs
+
+  def installedFiles =
+    makePlan cmdline inputs
+    | setPlanResources buildEnv.getFreedomESDKBuildEnvResources
+    | setPlanFnOutputs fnOutputs
+    | setPlanLocalOnly True
+    | setPlanDirectory runDir
+    | runJob
+    | getJobOutputs
+
+  # Using the generated design.dts, run Freedom Metal code generation for
+  # the target and install the generated code in the BSP directory.
+  #
+  # This step is only necessary as long as Product wants the code generators
+  # to not be shipped with the delivery. If the code generators are provided,
+  # Freedom Metal should have code generation occur on a per-application basis,
+  # which will happen automatically if the generated Metal sources are not
+  # present in the BSP directory.
+  def generatedMetalSources =
+    def designDTS =
+      installedFiles
+      | filter (\f matches `.*design\.dts` f.getPathName)
+      | head
+      | getOrElse (makeBadPath (makeError "Unable to get generated design.dts"))
+    def coreDTS =
+      installedFiles
+      | filter (\f matches `.*core\.dts` f.getPathName)
+      | head
+      | getOrElse (makeBadPath (makeError "Unable to get generated core.dts"))
+    makeFreedomMetalCodeGenerationOptions designDTS (coreDTS, Nil) bspDir
+    | runFreedomMetalCodeGeneration sdkDir
+    | getJobOutputs
+
+  FreedomESDKTarget sdk targetName targetType (installedFiles ++ generatedMetalSources)
```

Comment:
Similar to my earlier comment, I think we should have this return a `Result FreedomESDKTarget Error` rather than letting `BadPath`s propagate. For this function in particular, I think the new `require` keyword will make this a lot easier to write code for, since before the `require` keyword, we needed to heavily nested pattern matches or pipelines.

I'm not sure if we have any good examples of `require` in the open source code, but I could point you to a few examples in internal code if you'd like.